client/state/invites/selectors.js

```diff
@@ -44,6 +44,22 @@ export function getNumberOfInvitesFoundForSite( state, siteId ) {
 	return state.invites.counts[ siteId ] || null;
 }
 
+/*
+ * Returns an invite object for the given site and invite ID, or false otherwise.
+ *
+ * TODO: searching the object can probably be done in a cleaner manner.
+ *
+ * @param  {Object}  state    Global state tree
+ * @param  {Number}  siteId   Site ID
+ * @param  {String}  inviteId Invite ID
+ * @return {Object}           Invite object
+ */
+export function getSelectedInvite( state, siteId, inviteId ) {
```

Comment:
It would be good to have somebody familiar with writing efficient selectors (in other words, somebody other than me!) review this selector to make sure that we are following best practices! /cc @nylen

Reply:
I think this selector needs caching (memoization).  See the [`createSelector` docs](https://github.com/Automattic/wp-calypso/blob/a70587157b33783ba4d5e0e4a1e136680d459d51/client/lib/create-selector/README.md#usage) for the standard way of doing this.

There's also been talk recently of writing tests to make sure the selector caching and invalidation work as intended.  Ping @samouri because I haven't seen this done before and I'm not sure how to do it.

Reply:
One important thing to test is that the selector is returning referentially equal objects given the same input.  One way you could test that would look like this:

```js
test('returns the same object given the same input', () => {
  const state = ...;
    
  const call1 = getSelectedInvite( state, siteId, inviteId );
  const call2 = getSelectedInvite( state, siteId, inviteId );
  // mocha
  expect( call1 ).equal( call2 );
  // jest 
  expect( call1 ).toBe( call2);
})
```

Reply:
the referential equality test ensures that we aren't causing needless React rerenders, but depending on how many invites we expect a site to have, it may make sense to either memoize this or modify the datastructure since its an O(n) lookup.

Rather than createSelector, this is a pretty good candidate for [treeSelect](https://github.com/Automattic/wp-calypso/tree/master/client/lib/tree-select) which you could use like this:
```jsx
export const getSelectedInvite = treeSelect(
  ( state, siteId ) => [state.invites.items[ siteId ] ],
  ( [ siteInvites ], siteId, inviteId ) => {
    return find( siteInvite, { key: inviteId } );
  }
);
```

The alternative would be to modify the reducer so that instead of an opaque list that you need to scan, it could be an object keyed by the inviteId so it would be an O(1) lookup

Reply:
See https://github.com/Automattic/wp-calypso/pull/22081 for other pending changes to this reducer.  The invites for a site will probably end up looking like `{ pending: [ ... ], accepted: [ ... ] }` because this matches the most common retrieval pattern.

I didn't know about `treeSelect`, thanks for the pointer!

Reply:
Thanks for all of the advice here. I've updated the selector and tests for the current PR. I think we'll continue to refactor the reducer in #22081.